packages/hdwallet-keepkey/src/ethereum.ts

```diff
@@ -188,21 +173,116 @@ export async function ethGetAddress(transport: Transport, msg: core.ETHGetAddres
 }
 
 export async function ethSignMessage(transport: Transport, msg: core.ETHSignMessage): Promise<core.ETHSignedMessage> {
-  const m = new Messages.EthereumSignMessage();
+  const m = new Ethereum.EthereumSignMessage();
   m.setAddressNList(msg.addressNList);
   m.setMessage(ethers.utils.isBytes(msg.message) ? ethers.utils.arrayify(msg.message) : toUTF8Array(msg.message));
   const response = await transport.call(Messages.MessageType.MESSAGETYPE_ETHEREUMSIGNMESSAGE, m, {
     msgTimeout: core.LONG_TIMEOUT,
   });
-  const sig = response.proto as Messages.EthereumMessageSignature;
+  const sig = response.proto as Ethereum.EthereumMessageSignature;
   return {
     address: eip55.encode("0x" + core.toHexString(sig.getAddress_asU8())), // FIXME: this should be done in the firmware
     signature: "0x" + core.toHexString(sig.getSignature_asU8()),
   };
 }
 
+export async function ethSignTypedData(
+  transport: Transport,
+  msg: core.ETHSignTypedData
+): Promise<core.ETHSignedTypedData> {
+  /**
+   * If the message to be signed is sufficiently small, the KeepKey can calculate the
+   * domain separator and message hashes. Otherwise, we need to pre-calculate hashes
+   * here and verify on device.
+   */
+
+  let result: Ethereum.EthereumTypedDataSignature;
+
+  const sTypes = JSON.stringify({ types: msg.types });
+  const sPrimaryType = JSON.stringify({ primaryType: msg.primaryType });
+  const sDomain = JSON.stringify({ domain: msg.domain });
+  const sMessage = JSON.stringify({ message: msg.message });
+
+  try {
+    if (sTypes.length > 2048 || sPrimaryType.length > 80 || sDomain.length > 2048 || sMessage.length > 2048) {
+      /* Pre-calculate domain separator and messages hashes and verify on KeepKey */
+      const hashableMessage: TypedData = {
+        domain: msg.domain,
+        types: msg.types,
+        primaryType: msg.primaryType,
+        message: msg.message,
+      };
+      let messageHash = undefined;
+      const domainSeparatorHash = getTypeHash(hashableMessage, "EIP712Domain");
+      const domainSeparatorHash64 = Buffer.from(domainSeparatorHash).toString("base64");
+
+      const t = new Ethereum.EthereumSignTypedHash();
+      t.setAddressNList(msg.addressNList);
+      t.setDomainSeparatorHash(domainSeparatorHash64);
+
+      if (msg.message) {
+        messageHash = getMessage(hashableMessage, true);
```

Comment:
message is required on `TypedData` and while testing, it looks like `getTypeHash` throws if the payload does not conform correctly as is, don't think we need this check and can get rid of the `let` above